marklogic-client-api/src/test/java/com/marklogic/client/test/datamovement/JSONSplitterTest.java

```diff
@@ -0,0 +1,241 @@
+            assertEquals(expectedArray[i], element);
+        }
+    }
+
+    @Test
+    public void testJSONSplitterMultiArray() throws Exception {
+
+        JSONSplitter splitter = JSONSplitter.makeArraySplitter();
+        FileInputStream fileInputStream = new FileInputStream(new File(jsonMultiArrayFile));
+        Stream<StringHandle> contentStream = splitter.split(fileInputStream);
+        assertNotNull(contentStream);
+
+        StringHandle[] result = contentStream.toArray(size -> new StringHandle[size]);
+        assertEquals(2, splitter.getCount());
+        assertNotNull(result);
+
+        String[] expected = new String[] {
+                "[{\"context\":[[{\"record\":\"first record\"},{\"record\":\"second record\"}]]}]",
+                "{\"record\":\"third record\"}"
+        };
+
+        for (int i = 0; i < result.length; i++) {
+            String element = result[i].get();
+            assertNotNull(element);
+            assertEquals(expected[i], element);
+        }
+    }
+
+    @Test
+    public void testJSONSplitterWrite() throws Exception {
+
+        JSONSplitter splitter = JSONSplitter.makeArraySplitter();
+        FileInputStream fileInputStream = new FileInputStream(new File(jsonArrayFile));
+        Stream<DocumentWriteOperation> contentStream = splitter.splitWriteOperations(new JsonFactory().createParser(fileInputStream));
+        assertNotNull(contentStream);
+
+        Iterator<DocumentWriteOperation> itr = contentStream.iterator();
+        int i = 0;
+        while (itr.hasNext()) {
+            DocumentWriteOperation docOp = itr.next();
+            String uri = docOp.getUri();
+            assertNotNull(docOp.getUri());
+
+            assertNotNull(docOp.getContent());
+            String docOpContent = docOp.getContent().toString();
+            assertEquals(docOpContent, expectedArray[i]);
+            i++;
+        }
+
+        assertEquals(5, splitter.getCount());
+
+    }
+
+    @Test
+    public void testCustomMultiArray() throws Exception {
+
+        MultiArrayVisitor visitor = new MultiArrayVisitor();
+        JSONSplitter splitter = new JSONSplitter(visitor);
+        FileInputStream fileInputStream = new FileInputStream(new File(jsonCustMultiArrayFile));
+        Stream<StringHandle> contentStream = splitter.split(fileInputStream);
+        assertNotNull(contentStream);
+
+        StringHandle[] result = contentStream.toArray(size -> new StringHandle[size]);
+        assertEquals(2, splitter.getCount());
+        assertNotNull(result);
+
+        String[] expected = new String[] {
+                "{\"record\":\"first record\"}",
+                "{\"record\":\"second record\"}"
+        };
+
+        for (int i = 0; i < result.length; i++) {
+            String element = result[i].get();
+            assertNotNull(element);
+            assertEquals(expected[i], element);
+        }
+    }
+
+    static public class MultiArrayVisitor extends JSONSplitter.Visitor {
```

Comment:
Should this test derive from ArrayVisitor? Should ArrayVisitor refactor the increment and decrement operations into public incrementArrayDepth() and decrementArrayDepth() methods as conveniences for subclasses that need to manage the array depth?

Reply:
Yes, that's better. 

I changed both my Visitor tests to extend ArrayVisitor. Added 3 public methods in ArrayVisitor, getArrayDepth(), incrementArrayDepth() and decrementArrayDepth(). Thank you!